src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultProjectWorkspaceStateGenerator.cs

```diff
@@ -134,6 +125,12 @@ private async Task UpdateWorkspaceStateAsync(Project workspaceProject, IProjectS
             return;
         }
 
+        // Specifically not using BeginBlock because we want to capture cases where tag helper discovery never finishes.
```

Comment:
Ideally we would add a `BeginBlock` with some way to report success/failure status. If you agree can you file an issue to add that and comment here with the issue number to resolve later?

Reply:
To me, this is pretty specific to this class, as its the use of the semaphore that is the key, as a bug in the code handling it could mean we never discover tag helpers. I think it would be odd for `BeginBlock` to fire an event at the start, which is what this needs. The success/fail might be interesting, but without a use case I'm not sure its worth an issue.